lib/emis/service.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+require 'common/client/base'
+require 'emis/configuration'
+require 'emis/messages/edipi_or_icn_message'
+require 'emis/responses'
+require 'common/client/middleware/request/soap_headers'
+require 'common/client/middleware/response/soap_parser'
+
+module EMIS
+  class Service < Common::Client::Base
+    def self.create_endpoints(endpoints)
+      endpoints.each do |endpoint|
+        operation = nil
+        request_name = nil
+        if endpoint.is_a?(Array)
+          request_name = endpoint[1]
+          operation = endpoint[0].to_s
+        else
+          operation = endpoint.to_s
+          request_name = "#{endpoint.to_s.camelize(:lower).sub(/^get/, '').camelize(:lower)}Request"
+        end
+        define_method(operation) do |edipi: nil, icn: nil|
+          make_request(
+            edipi: edipi,
+            icn: icn,
+            request_name: request_name,
+            operation: operation,
+            response_type: "EMIS::Responses::#{operation.camelize}Response".constantize
+          )
+        end
+      end
+    end
+
+    protected
+
+    def make_request(edipi: nil, icn: nil, response_type:, operation:, request_name:)
+      message = create_edipi_or_icn_message(
+        edipi: edipi,
+        icn: icn,
+        request_name: request_name
+      )
+      raw_response = perform(
+        :post,
+        '',
+        message,
+        soapaction: "http://viers.va.gov/cdi/eMIS/#{operation.camelize(:lower)}/v1"
+      )
+      response_type.new(raw_response)
+      # :nocov:
```

Comment:
This may need eMIS involvement but does their service return 'proper' http errors or can it also return XML defined service errors in 200 responses? (like MVI, EVSS do). Probably good idea to get some specs for invalid/error responses too.